src/Data/List/Lazy.purs

```diff
@@ -705,3 +706,11 @@ foldM f a xs =
          Nothing -> pure a
          Just { head: b, tail: bs } ->
                        f a b >>= \a' -> foldM f a' bs
+
+-- | Perform a right fold lazily
+foldr' :: forall a b. Z.Lazy b => (a -> b -> b) -> b -> List a -> b
```

Comment:
Honestly, given the definition of `foldr` as `foldl` (strict) on the reversed list, and the initial impetus for this change in #111 , I'm at a loss for how such a thing could be done.

Would love to know if you have something in mind.